log_receivers.go

```diff
@@ -0,0 +1,252 @@
+/*
+LogReceivers are structs that define various log event endpoints that can
+receive IMQSV8 log events. Any new receiver can be added by extending the
+LogReceiver struct and implementing the Send(messages []*logMsg) interface method.
+
+The initial design here was to decouple the main logscraper routine from the
+actual sending of the events, as delays/issues in the sending to a receiver would
+affect the main routine as well as timeous delivery to other receivers. However, seeing
+that we only have 2 receivers at this stage, and no concrete requirements around
+delivery time of events and retention, we'll be sending the events in the same routine.
+This can always be changed by reintroducing the usage of the LogEvents channel ( or perhaps
+sending the events to RabbitMQ), once we have firmer requirements.
+*/
+
+package logscraper
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+)
+
+var receivers = make(map[string]Relay)
+var datadogSeverities = map[string]bool{
+	"ERROR": true,
+	"E":     true,
+	"FATAL": true,
+	"F":     true,
+}
+
+type Relay interface {
+	Send(messages []*LogMsg)
+	Receive(messages []*LogMsg)
+}
+
+type LogReceiver struct {
+	s         *Scraper
+	URL       string
+	ApiKey    string
+	LogEvents chan []*LogMsg
+}
+
+type LogglyReceiver struct {
+	LogReceiver
+}
+
+type DatadogReceiver struct {
+	LogReceiver
+	Host string
+}
+
+type logglyJsonMsg struct {
+	Host             string  `json:"host"`
+	OwnHostname      string  `json:"ownhostname"`
+	Source           string  `json:"source"`
+	Time             string  `json:"timestamp"`
+	Severity         string  `json:"severity,omitempty"`
+	Message          string  `json:"message,omitempty"`
+	ProcessID        int64   `json:"process_id,omitempty"`
+	ThreadID         int64   `json:"thread_id,omitempty"`
+	ClientIP         string  `json:"client_ip,omitempty"`
+	Request          string  `json:"request,omitempty"`
+	ResponseCode     string  `json:"response_code,omitempty"`
+	ResponseBytes    int64   `json:"response_bytes,omitempty"`
+	ResponseDuration float64 `json:"response_duration,omitempty"`
+	JavaClass        string  `json:"java_class,omitempty"`
+}
+
+type datadogJsonMessage struct {
+	Host           string `json:"host"`
+	Title          string `json:"title"`
+	Text           string `json:"text"`
+	Time           int64  `json:"date_happened"`
+	Tags           string `json:"tags,omitempty"`
+	AlertType      string `json:"alert_type"`
+	AggregationKey string `json:"aggregation_key,omitempty"`
+}
+
+/*
+Encodes all messages into a single json payload to send to Loggly
+*/
+func (lr *LogglyReceiver) Send(messages []*LogMsg) {
+	output := &bytes.Buffer{}
+	encoder := json.NewEncoder(output)
+	for _, message := range messages {
+		message.toLogglyJson(encoder)
+	}
+	var resp *http.Response
```

Comment:
No need to declare resp here, the type definition is handled automatically in the next line